### DISCUSSION TITLE
A loader that returns the compiled source (with dependencies) as string

### DISCUSSION THREAD

**Discussion by LukasBombach** [webpack/webpack#14984] [Q&A]
I am trying to write a loader the returns the compiled contents of a file as a string. So if I do

```ts
const sourceCode: string = require('compiled-loader!./myFile.js');
```

I would want to get the compiled sourcode—including bundled imports—as string, so if `myFile.js` is

```ts
import { log } from "./logger";
log("hello world");

//   ↓   I want the code returned from the loader to include 'logger.js'

function log(msg) { console.log(msg); }
log("hello world");
```

For this I would spawn a childCompiler, but I am not sure if I am doing this right and I am already getting warnings regarding caching and manifest outputs.

```js
const { fs: memfs } = require("memfs");

module.exports = function compiledLoader(content, map, meta) {
  const COMPILER_NAME = "Compiled Loader Compiler";
  const EntryOptionPlugin = this._compiler.webpack.EntryOptionPlugin;
  const outputOptions = { filename: "output.js" };
  const loaderCallback = this.async();

  // Prevent recursion
  if (this._compiler.name === COMPILER_NAME) {
    loaderCallback(null, content, map, meta);
    return;
  }

  // Create the child compiler
  const childCompiler = this._compilation.createChildCompiler(COMPILER_NAME, outputOptions);
  childCompiler.outputFileSystem = memfs;

  // Use the requested file as entry point
  EntryOptionPlugin.applyEntryOption(childCompiler, this._compiler.context, {
    child: { import: [this.resourcePath] },
  });

  // Compile the file with its dependencies
  childCompiler.runAsChild((_error, _entries, childCompilation) => {
    const source = childCompilation.assets[outputOptions.filename]?.source();

    // For brievity I am omitting error handling here

    loaderCallback(null, `module.exports = ${JSON.stringify(source)};`);
  });
};
```

I wonder if webpack has a built-in solution for this or if there are other easier options

**Comment by alexander-akait** [comment]
You can use `importModule`, example https://github.com/webpack-contrib/mini-css-extract-plugin/blob/master/src/loader.js#L215 and avoid child compilations

**Reply by LukasBombach** [reply]
This seems to return a module object and not a string

**Reply by alexander-akait** [reply]
Adding entries inside loader can potentially create deadlock

**Reply by LukasBombach** [reply]
oh, so, is there an alternative?

**Reply by alexander-akait** [reply]
Using `importModule`, just export it from module

**Reply by LukasBombach** [reply]
Sorry, I don't seem to understand how I can get the results from `importModule` as string.

If i do `const result = await importModule("./something.ts");`

the result is an object. If is stringify that, that will be `[object Object]`. If I stringify its exports individiually, there will probably tons of errors, like, not being able to stringify `undefined`

**Reply by alexander-akait** [reply]
What do you want achieve, can you provide small example?

**Reply by LukasBombach** [reply]
I created a next js application here: https://github.com/LukasBombach/compile-to-string-loader-example

What I am trying to do is this:

I wrote two loaders: 

the `cheap-replace-loader` finds the function `compileToString` in any file and replaces

```diff
- compileToString("./path/to/any/file")
+ compileToString(require("compile-to-string-loader.js!./path/to/any/file"))
```

the `compile-to-string-loader.js` is supposed to take `./path/to/any/file`, compile it with all its dependencies (bundle it essentially) and return its contents as string, so like

```js
module.exports = async function compiletoStringLoader() {
  const callback = this.async();
  const result = await this.importModule(this.resourcePath);  
  callback(null, `module.exports = ${JSON.stringify(result)}`);
};
```

( which of course does not work like that), but the idea is that I have a loader that compiles the requested file and returns its compiled source as string.

Or rather, in the end, the idea is to have a function `compileToString` that takes a path to a file as an argrument and will return the compiled sources of the file as a string

**Reply by LukasBombach** [reply]
I created a branch that uses a child compiler instead:

https://github.com/LukasBombach/compile-to-string-loader-example/tree/with-a-child-compiler

This method works!

This is the `compile-to-string-loader.js` with a child compiler

https://github.com/LukasBombach/compile-to-string-loader-example/blob/with-a-child-compiler/compile-to-string-loader.js
